### ISSUE TITLE
Prevent future deadlocks with an asynchronous actor model

### ISSUE THREAD

**Issue by Arceliar** [yggdrasil-network/yggdrasil-go#518]
tl;dr we often make the code deadlock when making changes to anything, but we could switch to a version of the actor model paradigm that doesn't deadlock by default the way cycles of goroutines and channels do.

Currently, we pass through quite a few goroutines to send or receive a packet, roughly something like the following:

|step | send | recv
|-----|------|-----
|1. | tun/tap reader | link reader
|2. | tunConn writer | switch
|3. | session send worker + crypto pool | router main goroutine
|4. | router write helper | session read worker + crypto pool
|5. | switch | tunConn reader
|6. | link writer | tun writer

It's very common for us to make changes to the code somewhere in these paths and get a deadlock as a result. That's something we need to keep in mind any time we have a network of goroutines communicating synchronously over channels.

There are only about 3 ways I can think of to prevent deadlocks:
1. Send into an infinitely large buffer (usually sending to a goroutine that does nothing but manage a slice buffer).
2. Drop messages, such as using `select` with a default case instead of a blocking send, or dropping more intelligently from a helper (similar to the above case).
3. Backpressure: the receiver signals that it's overloaded, and the sender stops trying to send until it's notified that the receiver is ready for more. This can theoretically propagate backwards to earlier steps, if the sender was a middle step in some pipeline. It's possible to do something like this with buffered channels and default cases in a `select` statement, but `select` doesn't compose well, so it basically needs to be re-implemented every time its used.

Since the synchronization in our pipeline causes more problems than it fixes at this point, we could consider switching to an asynchronous (but still ordered) message passing system based on the actor model. I've thrown together a [small actor model framework](https://github.com/Arceliar/phony) as a starting point, which takes the backpressure approach. This means we'd be able to pass packets along the pipeline without any infinite buffers or unintentional packet drops (only the intentional ones in the switch). The one thing to be careful about is to not let actors take blocking actions, but those are edge cases (tuntap and TCP read/writes, and possibly `yggdrasil.Conn`-related things, where we need to be a little bit careful about making the actors interface with non-actor goroutines and/or blocking syscalls).

**Comment by Arceliar** [comment]
Most of the work here was done in #519, so closing.
